### PR TITLE
planner: fix `Merge` hint for view

### LINKED ISSUE #36666: "merge"  hint in CTE with view will not work
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

 create table t(a int);
insert into t values (1);
create view v as select * from t;
explain with cte as (select /*+merge()*/ * from v) select * from cte;

### 2. What did you expect to see? (Required)
merge in cte works
### 3. What did you see instead (Required)
```sql
mysql> explain with cte as (select * from v) select * from cte;
+--------------------------------+---------+-----------+---------------+--------------------------------+
| id                             | estRows | task      | access object | operator info                  |
+--------------------------------+---------+-----------+---------------+--------------------------------+
| CTEFullScan_11                 | 1.00    | root      | CTE:cte       | data:CTE_0                     |
| CTE_0                          | 1.00    | root      |               | Non-Recursive CTE              |
| └─TableReader_9(Seed Part)     | 1.00    | root      |               | data:TableFullScan_8           |
|   └─TableFullScan_8            | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+--------------------------------+---------+-----------+---------------+--------------------------------+
4 rows in set (0.00 sec)

mysql> explain with cte as (select /*+merge()*/ * from v) select * from cte;
+--------------------------------+---------+-----------+---------------+--------------------------------+
| id                             | estRows | task      | access object | operator info                  |
+--------------------------------+---------+-----------+---------------+--------------------------------+
| CTEFullScan_11                 | 1.00    | root      | CTE:cte       | data:CTE_0                     |
| CTE_0                          | 1.00    | root      |               | Non-Recursive CTE              |
| └─TableReader_9(Seed Part)     | 1.00    | root      |               | data:TableFullScan_8           |
|   └─TableFullScan_8            | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+--------------------------------+---------+-----------+---------------+--------------------------------+
4 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

| Release Version: v6.2.0
Edition: Community
Git Commit Hash: 65f7e2e8b71ea7d312cfb35f87e6039271cb2dd8
Git Branch: heads/refs/tags/v6.2.0
UTC Build Time: 2022-07-27 10:10:50
GoVersion: go1.18.2

### PR DESCRIPTION
**Pull request by dayicklp**
### What problem does this PR solve?

Issue Number: close #36666 

Problem Summary:
When the CTE query has view, hint will not work. Because the current CTE's hint information will be overwritten in BuildView, so we save the CTE hint information and pass it back after BuildView.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by dayicklp** [comment]
The subquery in CTE which has `Megre` hint also has this issue, but this subquery will not go into buildSubquery, I need find a place to close it.

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -4833,6 +4833,14 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
...
 	originalVisitInfo := b.visitInfo
 	b.visitInfo = make([]visitInfo, 0)
+	saveCte := make([]*cteInfo, len(b.outerCTEs))
+	copy(saveCte, b.outerCTEs)
+	o := b.buildingCTE
+	b.buildingCTE = false
+	defer func() {
+		b.outerCTEs = append(saveCte, b.outerCTEs...)
+		b.buildingCTE = o
+	}()
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:4843]: Add comments and tests for this part.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -4833,6 +4837,23 @@ func (b *PlanBuilder) BuildDataSourceFromView(ctx context.Context, dbName model.
 	}
 	originalVisitInfo := b.visitInfo
 	b.visitInfo = make([]visitInfo, 0)
+
+	//For the case that views appear in CTE queries,
+	//we need to save the CTEs after the views are established.
+	var saveCte []*cteInfo
+	if b.outerCTEs != nil {
```
Comment (wjhuang2016) [inline planner/core/logical_plan_builder.go:4844]: len(b.outerCTEs) > 0

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -6693,7 +6714,8 @@ func (b *PlanBuilder) buildCte(ctx context.Context, cte *ast.CommonTableExpressi
 		}
 		b.buildingRecursivePartForCTE = saveCheck
 	} else {
-		p, err = b.buildResultSetNode(ctx, cte.Query.Query)
+		//When the query is CTE, we will turn the var to true.
```
Comment (wjhuang2016) [inline planner/core/logical_plan_builder.go:6717]: The comment is meaningless, you can remove it.

File: planner/core/planbuilder.go [hunk 4]
```diff
@@ -509,6 +509,7 @@ type PlanBuilder struct {
 	allocIDForCTEStorage        int
 	buildingRecursivePartForCTE bool
 	buildingCTE                 bool
+	isCTE                       bool
```
Comment (wjhuang2016) [inline planner/core/planbuilder.go:512]: Please add a comment for it

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -3797,9 +3799,16 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 	}
 
 	// Determines whether to use the Merge hint in a CTE query.
```
Comment (wjhuang2016) [inline planner/core/logical_plan_builder.go:3801]: Please descript the rule here.

File: planner/core/logical_plan_builder.go (hunk 3 above)
Comment (dayicklp) [reply planner/core/logical_plan_builder.go:6717]: OK

File: planner/core/logical_plan_builder.go (hunk 5 above)
Comment (dayicklp) [reply planner/core/logical_plan_builder.go:3801]: Rules for determining Merge hint have been added

**Review by wjhuang2016 — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/logical_plan_builder.go [hunk 6]
```diff
@@ -6711,6 +6735,7 @@ func (b *PlanBuilder) buildCte(ctx context.Context, cte *ast.CommonTableExpressi
 
 // buildRecursiveCTE handles the with clause `with recursive xxx as xx`.
 func (b *PlanBuilder) buildRecursiveCTE(ctx context.Context, cte ast.ResultSetNode) error {
+	b.isCTE = true
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:6738]: we set true here. where should we set it to false?

Comment (dayicklp) [reply planner/core/logical_plan_builder.go:6738]: The default is `false` for non-CTE queries at building. We change this to `true` for recursive and non-recursive CTEs, but this does not affect subsequent queries.

**Review by winoros — APPROVED** [review]

**Pull request merged**